### ISSUE TITLE
How do I build for a specific version of minecraft?

### ISSUE THREAD

**Issue by xiaodown** [cabaletta/baritone#4574]
## What do you need help with?

Hey Leijurv.

Current version works with 1.21.3 - but if I want to build for 1.21.1, how can I do that?  (or somehow modify the build to force it to ignore the version restrictions)?

## Final checklist
- [x] I know how to properly use check boxes 
- [X] I have not used any OwO's or UwU's in this issue.

**Comment by xiaodown** [comment]
(I tried unpacking the jar and editing neoforge.mods.toml, then rezipping it, but that didn't work)

**Comment by PizzaGod873** [comment]
#4540

**Comment by ZacSharp** [comment]
There is a (unmaintained and slightly outdated) 1.20.1 branch you can build from. If you need more recent changes on 1.21.1 you'll need to look into the changes made when updating from 1.21.1 to 1.21.3 and revert them (locally).

**Comment by xiaodown** [comment]
Ok, yeah, there is actually a `1.21.1` branch (https://github.com/cabaletta/baritone/tree/1.21.1), and I've tried building it, but without much success - both using powershell and using an ubuntu VM.  But I'm only a gradle novice, so ...maybe i'll try the other branch.

**Comment by PizzaGod873** [comment]
> Ok, yeah, there is actually a `1.21.1` branch (https://github.com/cabaletta/baritone/tree/1.21.1), and I've tried building it, but without much success - both using powershell and using an ubuntu VM. But I'm only a gradle novice, so ...maybe i'll try the other branch.

Make sure you are using java 17

**Comment by xiaodown** [comment]
Gooooooood call, I've been trying with various versions of java.  Is there a preference between oracle / openjdk?

**Comment by PizzaGod873** [comment]
Use adoptium, and get x64 jdk. If that doesn’t work I’ll send you a download link.

**Comment by xiaodown** [comment]
Sigh, I wish I understood what I'm doing wrong.

```
* What went wrong:
A problem occurred configuring project ':forge'.
> Failed to notify project evaluation listener.
   > Cannot find a Java installation on your machine matching this tasks requirements: {languageVersion=21, vendor=any, implementation=vendor-specific} for LINUX on x86_64.
      > No locally installed toolchains match and toolchain download repositories have not been configured. 
```

that's very indicative of the problems I've been having.  And I'm using the adoptium java 17:

```
 # java --version
openjdk 17.0.13 2024-10-15
OpenJDK Runtime Environment Temurin-17.0.13+11 (build 17.0.13+11)
OpenJDK 64-Bit Server VM Temurin-17.0.13+11 (build 17.0.13+11, mixed mode, sharing) 
```

(and this is a fresh ubuntu 22.04LTS vm I spun up just to compile this).

**Comment by xiaodown** [comment]
I got it to compile using temurin jdk21 but it crashes minecraft.  Here was my process:

```
git clone https://github.com/cabaletta/baritone.git
apt-get update; apt-get dist-upgrade
apt-get install -y wget apt-transport-https gnupg
wget -O - https://packages.adoptium.net/artifactory/api/gpg/key/public | apt-key add -
echo "deb https://packages.adoptium.net/artifactory/deb $(awk -F= '/^VERSION_CODENAME/{print$2}' /etc/os-release) main" | tee /etc/apt/sources.list.d/adoptium.list
apt-get update
apt-get install temurin-21-jdk
java --version
cd baritone/
git fetch; git switch 1.21.1
./gradlew build
```

That succeeds (fresh ubuntu 22.04) but the output jar doesn't work :(.

**Comment by xiaodown** [comment]
Build also succeeds for me with the 1.21 branch + temurin jdk21, and editing the neoforge meta to support 1.21.1, but still crashes blockgame :/.  Sigh, I feel so dumb.

**Comment by PizzaGod873** [comment]
I’ll send the link when I’m home

**Comment by xiaodown** [comment]
Thank you so much, man.  I apologize for being helpless.  I wish I knew more and didn't have to bug kind strangers.

**Comment by PizzaGod873** [comment]
Oh you got it finished?

**Comment by xiaodown** [comment]
no, sorry.  I am able to get the gradle build to work and output jars, using various branches and various versions of java, but every one i've managed to build crashes minecraft.

**Comment by xiaodown** [comment]
@PizzaGod873 If you do have a 1.21.1 jar that works, I'd really love a download link.  Apologies for bugging you, but I have had no luck.

**Comment by ZacSharp** [comment]
You can also enable GitHub Actions on your repository and trigger the build workflow on the branch you want by pushing a dummy commit.

**Comment by ZacSharp** [comment]
Just built this set of jars from the last commit on the 1.21.2 branch [baritone-1.10.2-11-g7e5dedd8f.zip](https://github.com/user-attachments/files/19000603/baritone-1.10.2-11-g7e5dedd8f.zip).
